Client/Frontend/Browser/BackForwardListViewController.swift

```diff
@@ -75,40 +70,52 @@ class BackForwardListViewController: UIViewController, UITableViewDataSource, UI
         super.init(nibName: nil, bundle: nil)
         
         loadSites(backForwardList)
+        loadSitesFromProfile()
     }
-    
-    func loadSites(backForwardList: WKBackForwardList) {
-        let sql = profile.favicons as! SQLiteHistory
-        var urls: [String] = [String]()
-        
-        for page in backForwardList.forwardList.reverse() {
-            urls.append(page.URL.absoluteString!)
-            listData.append((page, .Forward))
-        }
-        
-        if let currentPage = backForwardList.currentItem {
-            currentRow = listData.count
-            urls.append(currentPage.URL.absoluteString!)
-            listData.append((currentPage, .Current))
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.addSubview(shadow)
+        view.addSubview(tableView)
+        self.snappedToBottom = self.bvc?.toolbar != nil
+        tableView.snp_makeConstraints { make in
+            make.height.equalTo(0)
+            make.left.right.equalTo(self.view)
         }
-        for page in backForwardList.backList.reverse() {
-            urls.append(page.URL.absoluteString!)
-            listData.append((page, .Backward))
+        shadow.snp_makeConstraints { make in
+            make.left.right.equalTo(self.view)
         }
-        
-        listData = listData.filter { !(($0.item.title ?? "").isEmpty && $0.item.URL.baseDomain()?.contains("localhost") ?? false)}
-        
+        remakeVerticalConstraints()
+        view.layoutIfNeeded()
+        scrollTableViewToIndex(currentRow)
+        setupDismissTap()
+    }
+
+    func loadSitesFromProfile() {
+        let sql = profile.favicons as! SQLiteHistory
+        let urls = self.listData.flatMap {$0.URL.isLocal ? $0.URL.getQuery()["url"]?.unescape() : $0.URL.absoluteString}
+
         sql.getSitesForURLs(urls).uponQueue(dispatch_get_main_queue()) { result in
-            if let cursor = result.successValue {
-                for cursorSite in cursor {
-                    if let site = cursorSite, let url = site?.url {
-                        self.sites[url] = site
-                    }
-                }
-                self.tableView.reloadData()
+            guard let results = result.successValue else {
+                return
             }
+            // Add all results into the sites dictionary
+            results.flatMap({$0}).forEach({site in
+                if let url = site?.url {
+                    self.sites[url] = site
+                }
+            })
+            self.tableView.reloadData()
         }
     }
+
+    func loadSites(bfList: WKBackForwardList) {
+        let items = bfList.forwardList.reverse() + [bfList.currentItem].flatMap({$0}) + bfList.backList.reverse()
+        self.currentItem = bfList.currentItem
+
+        //error url's are OK as they are used to populate history on session restore.
+        listData = items.filter({ return !($0.URL.isLocal && $0.URL.path != "/errors/error.html")})
```

Comment:
Why filter sites here at all? Shouldn't the back/forward UI match exactly what would happen if we were to use the back/forward buttons?

Rather than showing the full error page URLs, though, we should extract the actual page that the error page is pointing to (using `ErrorPageHelper.originalURLFromQuery()`). If we show reader mode URLs in this list (which I think we should), we should do the same using `ReaderModeUtils.decodeURL()`.